File: libecole/benchmarks/src/scip/bench-model.cpp
#include <benchmark/benchmark.h>
#include <scip/scip.h>

#include "index-branchrule.hpp"
#include "utils.hpp"

using namespace ecole;

template <typename Func> auto benchmark_solve(benchmark::State& state, Func&& func_to_bench) {
	auto n_nodes = ecole::scip::long_int{0};
	auto n_lp_iterations = ecole::scip::long_int{0};

	for (auto _ : state) {
		state.PauseTiming();
		auto model = get_model();

		state.ResumeTiming();
		func_to_bench(model);
		state.PauseTiming();

		n_nodes += SCIPgetNTotalNodes(model.get_scip_ptr());
		n_lp_iterations += SCIPgetNLPIterations(model.get_scip_ptr());
	}

	using benchmark::Counter;
	state.counters["Nodes"] = Counter{static_cast<double>(n_nodes), Counter::kAvgIterations};
	state.counters["LP Iterations"] = Counter{static_cast<double>(n_lp_iterations), Counter::kAvgIterations};
}

auto presolve(benchmark::State& state) {
	benchmark_solve(state, [](auto& model) { model.presolve(); });
}
BENCHMARK(presolve)  // NOLINT(cert-err58-cpp)
	->MeasureProcessCPUTime()
	->UseRealTime()
	->Unit(benchmark::kMillisecond);

auto ecole_reverse_control(benchmark::State& state) {
	benchmark_solve(state, [](auto& model) {
		model.solve_iter();
		while (!model.solve_iter_is_done()) {
			model.solve_iter_branch(model.lp_branch_cands()[0]);
		}
	});
}
BENCHMARK(ecole_reverse_control)  // NOLINT(cert-err58-cpp)
	->MeasureProcessCPUTime()
	->UseRealTime()
	->Unit(benchmark::kMillisecond);

auto scip_branch_rule(benchmark::State& state) {
	benchmark_solve(state, [](auto& model) {
		auto branch_rule = std::make_unique<ecole::scip::IndexBranchrule>(model.get_scip_ptr(), "FirstVarBranching", 0UL);
		SCIPincludeObjBranchrule(model.get_scip_ptr(), branch_rule.release(), true);
		model.solve();
	});
}
BENCHMARK(scip_branch_rule)  // NOLINT(cert-err58-cpp)
	->MeasureProcessCPUTime()
	->UseRealTime()
	->Unit(benchmark::kMillisecond);

File: python/benchmarks/bench_model.py
import google_benchmark as benchmark
import pyscipopt

from index_branchrule import IndexBranchrule
from utils import get_model


def benchmark_model_solving(func_to_bench):
    @benchmark.register(name=func_to_bench.__name__)
    @benchmark.option.measure_process_cpu_time()
    @benchmark.option.use_real_time()
    @benchmark.option.unit(benchmark.kMillisecond)
    def _benchmark(state):
        n_nodes = 0
        n_lp_iterations = 0
        while state:
            state.pause_timing()
            model = get_model()

            state.resume_timing()
            func_to_bench(model)
            state.pause_timing()

            n_nodes += model.as_pyscipopt().getNNodes()  # FIXME different from C++
            n_lp_iterations += model.as_pyscipopt().getNLPIterations()  # FIXME Not all nodes
            state.resume_timing()

        state.counters["Nodes"] = benchmark.Counter(n_nodes, benchmark.Counter.kAvgIterations)
        state.counters["LP Iterations"] = benchmark.Counter(
            n_lp_iterations, benchmark.Counter.kAvgIterations
        )

    return _benchmark


@benchmark_model_solving
def presolve(model):
    model.presolve()


@benchmark_model_solving
def ecole_reverse_control(model):
    model.solve_iter()
    while not model.solve_iter_is_done():
        model.solve_iter_branch(model.lp_branch_cands[0])


@benchmark_model_solving
def pyscipopt_branchrule(model):
    pyscipopt_model = model.as_pyscipopt()
    pyscipopt_model.includeBranchrule(
        IndexBranchrule(pyscipopt_model, 0),
        "IndexBranchrule",
        "Branch on first LP branch cand",
        priority=536870911,
        maxdepth=-1,
        maxbounddist=1.0,
    )
    # Call `SCIPsolve` from PySCIPOpt because Ecole releases the GIL which is invalid if the Branchrule is written in
    # Python.
    pyscipopt_model.optimize()


if __name__ == "__main__":
    benchmark.main()
